File: src/tools/help-center.js
import { z } from 'zod';
import { zendeskClient } from '../zendesk-client.js';

export const helpCenterTools = [
  {
    name: "list_articles",
    description: "List Help Center articles. Returns paginated results.",
    schema: {
      page: z.number().min(1).optional().describe("Page number for pagination"),
      per_page: z.number().min(1).max(100).optional().describe("Number of articles per page (1-100)"),
      sort_by: z.string().optional().describe("Field to sort by"),
      sort_order: z.enum(["asc", "desc"]).optional().describe("Sort order (asc or desc)")
    },
    annotations: { readOnlyHint: true, destructiveHint: false, idempotentHint: true },
    handler: async ({ page, per_page, sort_by, sort_order }) => {
      try {
        const params = {};
        if (page !== undefined) params.page = page;
        if (per_page !== undefined) params.per_page = per_page;
        if (sort_by !== undefined) params.sort_by = sort_by;
        if (sort_order !== undefined) params.sort_order = sort_order;
        const result = await zendeskClient.listArticles(params);
        return {
          content: [{ type: "text", text: JSON.stringify(result, null, 2) }]
        };
      } catch (error) {
        return {
          content: [{ type: "text", text: `Error listing articles: ${error.message}` }],
          isError: true
        };
      }
    }
  },
  {
    name: "get_article",
    description: "Get a specific Help Center article by ID, including body content.",
    schema: {
      id: z.number().describe("Article ID")
    },
    annotations: { readOnlyHint: true, destructiveHint: false, idempotentHint: true },
    handler: async ({ id }) => {
      try {
        const result = await zendeskClient.getArticle(id);
        return {
          content: [{ type: "text", text: JSON.stringify(result, null, 2) }]
        };
      } catch (error) {
        return {
          content: [{ type: "text", text: `Error getting article: ${error.message}` }],
          isError: true
        };
      }
    }
  },
  {
    name: "create_article",
    description: "Create a new Help Center article in a specified section.",
    schema: {
      title: z.string().max(500).describe("Article title"),
      body: z.string().max(1048576).describe("Article body content (HTML)"),
      section_id: z.number().describe("Section ID where the article will be created"),
      locale: z.string().max(10).optional().describe("Article locale (e.g., 'en-us')"),
      draft: z.boolean().optional().describe("Whether the article is a draft"),
      permission_group_id: z.number().optional().describe("Permission group ID for the article"),
      user_segment_id: z.number().optional().describe("User segment ID for the article"),
      label_names: z.array(z.string().max(100)).max(20).optional().describe("Labels for the article")
    },
    annotations: { readOnlyHint: false, destructiveHint: false, idempotentHint: false },
    handler: async ({ title, body, section_id, locale, draft, permission_group_id, user_segment_id, label_names }) => {
      try {
        const articleData = { title, body };
        if (locale !== undefined) articleData.locale = locale;
        if (draft !== undefined) articleData.draft = draft;
        if (permission_group_id !== undefined) articleData.permission_group_id = permission_group_id;
        articleData.user_segment_id = user_segment_id !== undefined ? user_segment_id : null;
        if (label_names !== undefined) articleData.label_names = label_names;
        const result = await zendeskClient.createArticle(articleData, section_id);
        return {
          content: [{ type: "text", text: `Article created successfully!\n\n${JSON.stringify(result, null, 2)}` }]
        };
      } catch (error) {
        return {
          content: [{ type: "text", text: `Error creating article: ${error.message}` }],
          isError: true
        };
      }
    }
  },
  {
    name: "update_article",
    description: "Update an existing Help Center article. Only provided fields will be changed.",
    schema: {
      id: z.number().describe("Article ID to update"),
      title: z.string().max(500).optional().describe("Updated article title"),
      body: z.string().max(1048576).optional().describe("Updated article body content (HTML)"),
      locale: z.string().max(10).optional().describe("Updated article locale (e.g., 'en-us')"),
      draft: z.boolean().optional().describe("Whether the article is a draft"),
      permission_group_id: z.number().optional().describe("Updated permission group ID"),
      user_segment_id: z.number().optional().describe("Updated user segment ID"),
      label_names: z.array(z.string().max(100)).max(20).optional().describe("Updated labels")
    },
    annotations: { readOnlyHint: false, destructiveHint: false, idempotentHint: true },
    handler: async ({ id, title, body, locale, draft, permission_group_id, user_segment_id, label_names }) => {
      try {
        const articleData = {};
        if (title !== undefined) articleData.title = title;
        if (body !== undefined) articleData.body = body;
        if (locale !== undefined) articleData.locale = locale;
        if (draft !== undefined) articleData.draft = draft;
        if (permission_group_id !== undefined) articleData.permission_group_id = permission_group_id;
        if (user_segment_id !== undefined) articleData.user_segment_id = user_segment_id;
        if (label_names !== undefined) articleData.label_names = label_names;

        const result = await zendeskClient.updateArticle(id, articleData);
        const article = result?.article || result?.translation || result;
        return {
          content: [{ type: "text", text: `Article updated successfully!\n\n${JSON.stringify({ article }, null, 2)}` }]
        };
      } catch (error) {
        return {
          content: [{ type: "text", text: `Error updating article: ${error.message}` }],
          isError: true
        };
      }
    }
  }
];

File: src/tools/help_center.py
"""MCP tool handlers for Zendesk Help Center articles."""

from __future__ import annotations

import json

from ..attribution import apply_article_attribution, get_current_entra_email
from ..zendesk_client import zendesk_client
from ..zendesk_user_resolver import resolve_zendesk_user_id


def _article_summary(a: dict) -> dict:
    return {
        "id": a.get("id"),
        "url": a.get("html_url") or zendesk_client.get_help_center_article_url(a["id"]),
        "title": a.get("title"),
        "section_id": a.get("section_id"),
        "locale": a.get("locale"),
        "draft": a.get("draft"),
        "created_at": a.get("created_at"),
        "updated_at": a.get("updated_at"),
        "author_id": a.get("author_id"),
        "label_names": a.get("label_names"),
    }


async def list_articles(
    page: int | None = None,
    per_page: int | None = None,
    sort_by: str | None = None,
    sort_order: str | None = None,
) -> str:
    """List Help Center articles with optional pagination and sorting."""
    params: dict = {}
    if page is not None:
        params["page"] = page
    if per_page is not None:
        params["per_page"] = per_page
    if sort_by is not None:
        params["sort_by"] = sort_by
    if sort_order is not None:
        params["sort_order"] = sort_order

    result = await zendesk_client.list_articles(params)
    articles = [_article_summary(a) for a in result.get("articles", [])]
    summary = {
        "count": result.get("count"),
        "next_page": result.get("next_page"),
        "articles": articles,
    }
    return json.dumps(summary, indent=2)


async def get_article(id: int) -> str:  # pylint: disable=redefined-builtin
    """Fetch a single Help Center article by ID, truncating the body to 2000 chars."""
    result = await zendesk_client.get_article(id)
    a = result.get("article") or result
    summary = {
        **_article_summary(a),
        "body": (a.get("body") or "")[:2000],
    }
    return json.dumps(summary, indent=2)


async def create_article(  # pylint: disable=too-many-arguments,too-many-positional-arguments
    title: str,
    body: str,
    section_id: int,
    locale: str | None = None,
    draft: bool | None = None,
    permission_group_id: int | None = None,
    user_segment_id: int | None = None,
    label_names: list[str] | None = None,
) -> str:
    """Create a new Help Center article in the specified section."""
    article_data: dict = {"title": title, "body": body}
    if locale is not None:
        article_data["locale"] = locale
    if draft is not None:
        article_data["draft"] = draft
    if permission_group_id is not None:
        article_data["permission_group_id"] = permission_group_id
    article_data["user_segment_id"] = (
        user_segment_id if user_segment_id is not None else None
    )
    if label_names is not None:
        article_data["label_names"] = label_names

    # Architecture F attribution: when the caller didn't provide author_id,
    # use the authenticated Entra user's Zendesk user_id. See ADR-015.
    # update_article intentionally does NOT inject — changing author_id on
    # update rewrites the article's author, not "last editor".
    user_email = get_current_entra_email()
    if user_email and "author_id" not in article_data:
        user_id = await resolve_zendesk_user_id(user_email)
        if user_id:
            apply_article_attribution(article_data, user_id)

    result = await zendesk_client.create_article(article_data, section_id)
    a = result.get("article") or result
    summary = _article_summary(a)
    return (
        f"Article #{a['id']} created successfully!\n\n{json.dumps(summary, indent=2)}"
    )


async def update_article(  # pylint: disable=redefined-builtin,too-many-arguments,too-many-positional-arguments
    id: int,
    title: str | None = None,
    body: str | None = None,
    locale: str | None = None,
    draft: bool | None = None,
    permission_group_id: int | None = None,
    user_segment_id: int | None = None,
    label_names: list[str] | None = None,
) -> str:
    """Update an existing Help Center article by ID."""
    article_data: dict = {}
    if title is not None:
        article_data["title"] = title
    if body is not None:
        article_data["body"] = body
    if locale is not None:
        article_data["locale"] = locale
    if draft is not None:
        article_data["draft"] = draft
    if permission_group_id is not None:
        article_data["permission_group_id"] = permission_group_id
    if user_segment_id is not None:
        article_data["user_segment_id"] = user_segment_id
    if label_names is not None:
        article_data["label_names"] = label_names

    result = await zendesk_client.update_article(id, article_data)
    a = result.get("article") or result.get("translation") or result
    summary = _article_summary(a)
    return f"Article #{id} updated successfully!\n\n{json.dumps(summary, indent=2)}"
